### PR TITLE
Support ANTLR 4.13.1 and modernize the project setup

### PR DESCRIPTION
**Pull request by lppedd**
This is a fairly big PR, so I recommend going through it commit by commit, each being self-descriptive.

What has been done:

1. Updated Kotlin to 1.9.21
2. Ditched the custom Gradle plugin
3. Updated Gradle to 8.3 and modernized the build scripts
4. Applied a bunch of currently opened PRs from this repo, and one PR from the ANTLR4 repo  
    #77  
    #75  
5. Solved a couple of issues  
    #81  
    #93  
    #106
6. Set up a skeleton to publish to a Maven repository
7. Restructured the Kotlin template (`Kotlin.stg`) to output working, beautified and readable code
8. Updated the runtime, commit by commit, to align to the Java runtime - version 4.13.1

## Testing

I've set up an additional module `antlr-kotlin-tests`, which is supposed to host all the multiplatform tests.  
I've only ported the `MiniCalc*` ones, but I don't think there are others to move.

## TO-DO

- [ ] Restore Travis (or another CI) build, and maybe Jitpack, although I'd like to avoid it and publish straight to Central
- [x] Update the README to explain how to setup the Kotlin target with Gradle's `antlr` plugin
- [x] ~Possibly restore example projects~ (or agree to move them to another repository - preferred IMO)
- [x] Possibly restore Native targets

### CODE REVIEW

**Comment by lppedd** [comment]
I encourage you to checkout the branch and experiment with it.

File: .travis.yml
Comment (ftomassetti) [inline .travis.yml:1]: I understand we are ditching Travis (good idea) to switch to GitHub Actions

File: antlr-kotlin-gradle-plugin/build.gradle.kts
Comment (ftomassetti) [inline antlr-kotlin-gradle-plugin/build.gradle.kts:1]: Are we just removing the gradle plugin?

File: antlr-kotlin-runtime/build.gradle.kts [hunk 1]
```diff
@@ -1,113 +1,57 @@
...
-        binaries {
-            staticLib()
-        }
+  jvm {
+    compilations.configureEach {
+      compilerOptions.configure {
+        jvmTarget.set(JvmTarget.JVM_1_8)
+        freeCompilerArgs.addAll(
+          "-Xjvm-default=all",
+          // TODO: verify the performance impact of null assertions
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/build.gradle.kts:23]: Perhaps we can remove these lines and open an issue instead

File: README.md [hunk 2]
```diff
@@ -15,6 +15,16 @@ Because we produce multi-platform code and a multi-platform runtime, the code wr
 
 The project should be considered experimental. Several parsers have been implemented and work so far using this target but not all the features are complete and well tested.
 
+The runtime and target's template are up-to-date with mainstream commit: `master/e9df464`
```
Comment (ftomassetti) [inline README.md:18]: @lppedd , could you also add a section about contributors specifying your name. You are showing quite a lot of love to this project and your contribution is making the difference, so I think your name should appear in the README

File: antlr-kotlin-runtime/build.gradle.kts (hunk 1 above)
Comment (lppedd) [reply antlr-kotlin-runtime/build.gradle.kts:23]: Yup those are just reminders. Will get rid of them before merging.  
I was discussing about assertions on Kotlin's Slack, and will open a YouTrack issue to get the same flags for the JS target.

File: build.gradle.kts [hunk 3]
```diff
@@ -1,55 +1,43 @@
...
-}
+  apply(plugin = "maven-publish")
 
-allprojects {
-    // ... because `version` is another var here.
-    // when version is hardcoded here, jitpack can not overwrite it.
-    // the default version can now be changed in gradle.properties
-    version = versionProperty
-    group = groupProperty
+  // TODO: setup a real Maven repository to publish to, e.g., Maven Central
```
Comment (ftomassetti) [inline build.gradle.kts:15]: We could publish on Maven Central, we have already an account to publish everything under the com.strumenta group

File: ci.sh
Comment (ftomassetti) [inline ci.sh:1]: Oh my, it was long overdue removing this!

File: jitpack.yml
Comment (ftomassetti) [inline jitpack.yml:1]: I will not miss jitpack :)

File: settings.gradle.kts [hunk 4]
```diff
@@ -1,7 +1,14 @@
+enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")
```
Comment (ftomassetti) [inline settings.gradle.kts:1]: I am not familiar with this

**Comment by ftomassetti** [comment]
Wow @lppedd what an amazing contribution: THANK YOU!

>  Restore Travis (or another CI) build, and maybe Jitpack, although I'd like to avoid it and publish straight to Central

I would suggest we use GitHub actions and we publish to Maven Central

>  Possibly restore example projects (or agree to move them to another repository - preferred IMO)

Let's move them to another project. We actually created the repository some time ago: https://github.com/strumenta/antlr-kotlin-examples

File: .travis.yml
Comment (lppedd) [reply .travis.yml:1]: Oh nice. Will you take a shot at it, or do you want me to do it?  
I'm a Jenkins guy so Actions would require experimentation most probably, and not sure how they'd work in a PR context.

File: antlr-kotlin-examples-js/.gitignore
[the remark this answers is not in the document]
Comment (lppedd) [reply antlr-kotlin-examples-js/.gitignore:1]: Example projects were cluttering the repository imo. Better move them to the place you pointed to in another comment.

File: antlr-kotlin-gradle-plugin/build.gradle.kts
Comment (lppedd) [reply antlr-kotlin-gradle-plugin/build.gradle.kts:1]: Yup, no more needed. It's easier to add the custom target to the classpath of the Gradle ANTLR plugin.
This plugin was mostly duplicating the Gradle one, so it was just a maintenance burden.

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[versions]
```
[the remark this answers is not in the document]
Comment (lppedd) [reply gradle/libs.versions.toml:1]: This is the best explanation of the Version Catalog.  
https://developer.android.com/build/migrate-to-catalogs

It's a recent addition, but extremely useful to centralize dependencies information.  
Plus, it lets you access dependencies in a typesafe way:
```
dependencies {
  implementation(libs.antlr4)
}
```
The above would refer to a catalog entry such as:
```
[versions]
antlr4 = "4.13.1"

[libraries]
antlr4 = { group = "org.antlr", name = "antlr4", version.ref = "antlr4" }
```

You can't do this with `gradle.properties`.

File: settings.gradle.kts (hunk 4 above)
Comment (lppedd) [reply settings.gradle.kts:1]: Typesafe accessors allow you to transform:
```
dependencies {
  implementation(project(":antlr-kotlin-target"))
}
```
into:
```
dependencies {
  implementation(projects.antlrKotlinTarget)
}
```

**Comment by lppedd** [comment]
Just pushed another couple of commits to add back the previously enabled Native targets.  
As with JS, Native targets lack proper implementations for `IdentityHashMap` and `WeakHashMap`.

`WeakHashMap` is tracked at https://youtrack.jetbrains.com/issue/KT-48075  
`IdentityMap` is tracked at https://youtrack.jetbrains.com/issue/KT-8435

Plus, native targets are **not** thread safe. I'll have to check what can be done with Kotlin's AtomicFU, but not now, maybe in some months.

**Comment by lppedd** [comment]
Regarding example projects, I'd say let's get this merged first (and published possibly), and then I can go back and update them in the other repo. I'll do some manual testing tho, to verify that they still work.

File: .travis.yml
Comment (ftomassetti) [reply .travis.yml:1]: Yes, I can do that as a separate thing. In essence if we want to run the gradle check task we could just create this file: https://github.com/Strumenta/kolasu/blob/master/.github/workflows/check.yml and be done with it

File: antlr-kotlin-examples-js/.gitignore
[the remark this answers is not in the document]
Comment (ftomassetti) [reply antlr-kotlin-examples-js/.gitignore:1]: Understood, good

File: antlr-kotlin-gradle-plugin/build.gradle.kts
Comment (ftomassetti) [reply antlr-kotlin-gradle-plugin/build.gradle.kts:1]: oh, great. We may just need to add some info about doing that in the Readme and be done with it

File: gradle/libs.versions.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ftomassetti) [reply gradle/libs.versions.toml:1]: oh, I see, thank you

File: settings.gradle.kts (hunk 4 above)
Comment (ftomassetti) [reply settings.gradle.kts:1]: oh, great!

File: README.md (hunk 2 above)
Comment (lppedd) [reply README.md:21]: Done!

**Comment by lppedd** [comment]
I've updated the readme, let me know if it's good enough.  
I think it's up to you now to add the publishing/signing bits and the Actions stuff.

I'll experiment with build + publication to see what's the best chain of commands, but we also need to consider we are building on two machines (at least).

**Comment by ftomassetti** [comment]
Ok, let's merge this one and then add GitHub actions and publishing

**Comment by lppedd** [comment]
Sounds good! Thanks!
I'll open another PR soon to address #104 and #38.

**Pull request merged**